### PR TITLE
Home task 9

### PR DESCRIPTION
**Pull request by AlexLeshyk**
Home task 9: Rxjs observables

### CODE REVIEW

File: src/app/core/components/header/header.component.ts [hunk 1]
```diff
@@ -8,26 +11,44 @@ import { UserEntity } from '../../models/user-entity.model';
 })
 export class HeaderComponent implements OnInit {
 
-  @Output() open = new EventEmitter<void>();
-  @Input() isAuth = true;
+  isAuth = true;
+  aSub: Subscription;
+  user : UserEntity;
+  // user : UserEntity = {
```
Comment (Lexani) [inline src/app/core/components/header/header.component.ts]: please remove comments

File: src/app/core/components/header/header.component.ts [hunk 2]
```diff
@@ -8,26 +11,44 @@ import { UserEntity } from '../../models/user-entity.model';
...
+    this.router.events.subscribe((event) => {
+      if (event instanceof NavigationEnd) {
+        this.updateId();
+      }
+    });
+  }
+
+  updateId(): void {
+    this.currentId = this.auth.getCurrentId();
+    if (this.currentId !== undefined) {
```
Comment (Lexani) [inline src/app/core/components/header/header.component.ts]: I prefer checking if(!this.currentId) // check on null , undefined, empty string. cause all such cases if occur, won't work further.

File: src/app/core/components/header/header.component.ts [hunk 3]
```diff
@@ -8,26 +11,44 @@ import { UserEntity } from '../../models/user-entity.model';
...
+      this.auth.getUserById(this.currentId).subscribe( item => {
+        this.user = item;
+      })
+    }
+  }
+
+  openPopup() {
+    // this.isAuth = false;
```
Comment (Lexani) [inline src/app/core/components/header/header.component.ts]: remove comment

File: src/app/core/components/login-page/login-page.component.ts [hunk 4]
```diff
@@ -21,16 +24,41 @@ export class LoginPageComponent implements OnInit {
   ngOnInit() {
   }
 
+  ngOnDestroy() {
+    if (this.aSub) {
```
Comment (Lexani) [inline src/app/core/components/login-page/login-page.component.ts]: Naming conversion for subscribtions is adding $ at the end of the name. You should write something like authentication$

File: src/app/course-page/components/add-course-page/add-course-page.component.ts [hunk 5]
```diff
@@ -10,52 +12,62 @@ import { Router } from '@angular/router';
...
+    private router: Router,
+    private loadService: LoadingService
   ) { }
 
   ngOnInit() {
     this.route.params.subscribe( (params: Params) => {
       this.isNewCourse = !params.id;
       if (!this.isNewCourse) {
-        // this.courseItem = this.itemCourseService.getItemById(+params.id);
+        // this.courseItem = this.itemCourseService.getItemById(params.id);
```
Comment (Lexani) [inline src/app/course-page/components/add-course-page/add-course-page.component.ts]: remove comment

File: src/app/course-page/components/course-item/course-item.component.ts [hunk 6]
```diff
@@ -31,13 +34,23 @@ export class CourseItemComponent implements OnInit, OnChanges {
...
+        console.log("This course won't be deleted", confirmation);
+      }
+    }
   }
 
   public updateCourseItem(): void {
-    if (this.auth.getAutorizationValue()) {
-      this.itemService.rememberId(this.courseItem.id);
-    }
+    // if (this.auth.getAutorizationValue()) {
```
Comment (Lexani) [inline src/app/course-page/components/course-item/course-item.component.ts]: remove comment

File: src/app/course-page/components/course-item/course-item.component.ts [hunk 7]
```diff
@@ -31,13 +34,23 @@ export class CourseItemComponent implements OnInit, OnChanges {
...
+    //   this.itemService.rememberId(this.courseItem.id);
+    // }
+
+    this.itemService.rememberId(this.courseItem.id);
+    this.loadService.loadingBlock();
```
Comment (Lexani) [inline src/app/course-page/components/course-item/course-item.component.ts]: method should have a verb in the name, cause method is an action. Please, rename it to smth like "showLoad"

File: src/app/course-page/components/course-item/course-item.component.ts [hunk 8]
```diff
@@ -31,13 +34,23 @@ export class CourseItemComponent implements OnInit, OnChanges {
...
 
   public updateCourseItem(): void {
-    if (this.auth.getAutorizationValue()) {
-      this.itemService.rememberId(this.courseItem.id);
-    }
+    // if (this.auth.getAutorizationValue()) {
+    //   this.itemService.rememberId(this.courseItem.id);
+    // }
+
+    this.itemService.rememberId(this.courseItem.id);
```
Comment (Lexani) [inline src/app/course-page/components/course-item/course-item.component.ts:49]: why do you need to save id here?

File: src/app/course-page/components/course-page/course-page.component.ts [hunk 9]
```diff
@@ -1,62 +1,136 @@
...
+import { SubscriptionLike, Subject } from 'rxjs';
+import { Router, ActivatedRoute, Params } from '@angular/router';
+import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
 
 @Component({
   selector: 'app-course-page',
   templateUrl: './course-page.component.html',
   styleUrls: ['./course-page.component.scss'],
-  changeDetection: ChangeDetectionStrategy.OnPush
+  // changeDetection: ChangeDetectionStrategy.OnPush
```
Comment (Lexani) [inline src/app/course-page/components/course-page/course-page.component.ts]: remove comment

File: src/app/course-page/components/course-page/course-page.component.ts [hunk 10]
```diff
@@ -1,62 +1,136 @@
...
   public change(): void {
     this.counter = this.counter + 1;
     console.log(this.counter);
-    // this.isShowCourse = this.isShowCourse ? false : true;
   }
 
-  public onItemDelete(item: CourseItem): void {
-    this.itemCourseService.deleteItem(item);
+  public onItemDelete(item: CourseItem) {
+    // this.itemCourseService.deleteItem(item);
```
Comment (Lexani) [inline src/app/course-page/components/course-page/course-page.component.ts]: remove comment

File: src/app/course-page/components/course-page/course-page.component.ts [hunk 11]
```diff
@@ -1,62 +1,136 @@
...
+    this.subscriptions.push(this.itemCourseService.removeItem(item)
+    .subscribe( () => {
+      this.courseItems = this.courseItems.filter( t => t.id !==  item.id);
```
Comment (Lexani) [inline src/app/course-page/components/course-page/course-page.component.ts]: do you need this code? After the deletion you call fetchItems and the updated list of items will be shown

File: src/app/course-page/components/course-page/course-page.component.ts [hunk 12]
```diff
@@ -1,62 +1,136 @@
...
+    }, error =>{
+      this.error = error.message;
+    }
+  ));
+  }
+
+  onSearch() {
+  }
+
+  onShowSecondRow() {
```
Comment (Lexani) [inline src/app/course-page/components/course-page/course-page.component.ts]: You definitely don't need four methods that are different only by parameters. Please change your method fetchItems to accept startIndex and count - parameters, and remove these separate methods.

File: src/app/course-page/components/course-page/course-page.component.ts [hunk 13]
```diff
@@ -1,62 +1,136 @@
...
+    this.fetchItems();
+  }
+
+  onShowAllItems() {
+    this.showParamsService.showAllItems();
+    this.fetchItems();
   }
 
   public ngOnChanges(changes: SimpleChanges): void {
     // console.log('OnChanges CoursePage Component', changes);
```
Comment (Lexani) [inline src/app/course-page/components/course-page/course-page.component.ts]: remove empty method

File: src/app/course-page/components/search-control/search-control.component.html [hunk 14]
```diff
@@ -1,4 +1,5 @@
 <div class="search-block-wrapper">
-  <input type="text" [ngModel]="inputValue" (ngModelChange)="onChange($event)"/>
-  <button class="btn primary-btn" (click)="onClick()">Search</button>
+  <!-- <input type="text" [ngModel]="inputValue" (ngModelChange)="onChange($event)"/> -->
```
Comment (Lexani) [inline src/app/course-page/components/search-control/search-control.component.html]: remove comments

File: src/app/course-page/services/item-course.service.ts [hunk 15]
```diff
@@ -1,96 +1,73 @@
...
+  public count = '5';
 
-  constructor() { }
+  constructor(private http: HttpClient) { }
 
   // Get list
-  public getItems(): CourseItem[] {
-    return this.items;
+  getItems(): Observable<CourseItem[]> {
+    let params = new HttpParams();
```
Comment (Lexani) [inline src/app/course-page/services/item-course.service.ts]: I suggest using the following  approach: this.http.get('http://localhost:888.....', { headers: headers, search: params })
Just pass parameters with their names and dont create separate new HttpParams.

File: src/app/course-page/services/item-course.service.ts [hunk 16]
```diff
@@ -1,96 +1,73 @@
...
+        console.log('Error: ', error.message)
+        return throwError(error)
+      })
+    )
   }
 
   // Get item by Id
   getItemById(id: number) {
-    return this.items.find((course: CourseItem) => course.id === id);
+    // return this.items.find((course: CourseItem) => course.id === id);
```
Comment (Lexani) [inline src/app/course-page/services/item-course.service.ts]: remove comment

File: src/app/course-page/services/item-course.service.ts [hunk 17]
```diff
@@ -1,96 +1,73 @@
...
+    let params = new HttpParams();
+    params = params.append('start', this.start);
+    params = params.append('count', this.count);
+    return this.http.get<CourseItem[]>('http://localhost:3004/courses', {
+      params,
+      observe: 'response'
+    })
+    .pipe(
+      map(response => {
+        // console.log('Response', response)
```
Comment (Lexani) [inline src/app/course-page/services/item-course.service.ts]: remove comment

File: src/app/course-page/services/show-params.service.ts [hunk 18]
```diff
@@ -0,0 +1,34 @@
+import { Injectable } from '@angular/core';
+import { ItemCourseService } from './item-course.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class ShowParamsService {
```
Comment (Lexani) [inline src/app/course-page/services/show-params.service.ts]: You don't need this class, please change fetchItems to be able to accept startIndex and count parameters.

File: src/app/shared/auth.interceptor.ts [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+@Injectable()
+export class AuthInterceptor implements HttpInterceptor {
+  constructor(
+    private auth: AuthorizationService,
+    private router: Router
+  ) {
+  }
+  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    // console.log('Intercept request', req.params);
```
Comment (Lexani) [inline src/app/shared/auth.interceptor.ts]: remove comment

**Review by Lexani — CHANGES_REQUESTED** [review]

File: src/app/core/components/header/header.component.ts (hunk 1 above)
Comment (AlexLeshyk) [reply src/app/core/components/header/header.component.ts]: removed

File: src/app/core/components/header/header.component.ts (hunk 2 above)
Comment (AlexLeshyk) [reply src/app/core/components/header/header.component.ts]: changed code

File: src/app/core/components/header/header.component.ts (hunk 3 above)
Comment (AlexLeshyk) [reply src/app/core/components/header/header.component.ts]: removed

File: src/app/core/components/login-page/login-page.component.ts (hunk 4 above)
Comment (AlexLeshyk) [reply src/app/core/components/login-page/login-page.component.ts]: changed naming

File: src/app/course-page/components/add-course-page/add-course-page.component.ts (hunk 5 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/add-course-page/add-course-page.component.ts]: removed

File: src/app/course-page/components/course-item/course-item.component.ts (hunk 6 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-item/course-item.component.ts]: removed

File: src/app/course-page/components/course-item/course-item.component.ts (hunk 7 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-item/course-item.component.ts]: changed naming of the method

File: src/app/course-page/components/course-page/course-page.component.ts (hunk 9 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-page/course-page.component.ts]: removed

File: src/app/course-page/components/course-page/course-page.component.ts (hunk 10 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-page/course-page.component.ts]: removed

File: src/app/course-page/components/course-page/course-page.component.ts (hunk 11 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-page/course-page.component.ts]: removed this part of code

File: src/app/course-page/components/course-page/course-page.component.ts (hunk 13 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-page/course-page.component.ts]: removed empty method

File: src/app/course-page/components/search-control/search-control.component.html (hunk 14 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/search-control/search-control.component.html]: removed comments in component code

File: src/app/shared/auth.interceptor.ts (hunk 19 above)
Comment (AlexLeshyk) [reply src/app/shared/auth.interceptor.ts]: removed

File: src/app/course-page/services/item-course.service.ts (hunk 16 above)
Comment (AlexLeshyk) [reply src/app/course-page/services/item-course.service.ts]: removed

File: src/app/course-page/services/item-course.service.ts (hunk 17 above)
Comment (AlexLeshyk) [reply src/app/course-page/services/item-course.service.ts]: removed

File: src/app/course-page/services/show-params.service.ts (hunk 18 above)
Comment (AlexLeshyk) [reply src/app/course-page/services/show-params.service.ts]: Removed this service

File: src/app/course-page/services/item-course.service.ts (hunk 15 above)
Comment (AlexLeshyk) [reply src/app/course-page/services/item-course.service.ts]: set params like without new HttpParams
      params: {
        'start': startIndex,
        'count': count
      }

File: src/app/course-page/components/course-page/course-page.component.ts (hunk 12 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-page/course-page.component.ts]: Removed unnecessary methods

File: src/app/course-page/components/course-item/course-item.component.ts (hunk 8 above)
Comment (AlexLeshyk) [reply src/app/course-page/components/course-item/course-item.component.ts:49]: I didn't know how to get id of current item for Breadcrumbs component/ So I did get & set methods

**Pull request merged**